File: go/bubblesort.go
package main

import "fmt"

var arr = [10]int{9, 3, 2, 8, 7, 5, 0, 1, 6, 4}

func main() {
	fmt.Printf("Sorting Collection %v\n", arr)
	for i := len(arr) - 1; i > 1; i-- {
		fmt.Printf("Iterating Outer over %v \n", i)
		for j := 0; j < i; j++ {
			fmt.Printf("Iterating Inner over %v \n", j)
			fmt.Printf("Comparing left %v - %v right %v - %v \n", arr[j], j, arr[j+1], j+1)
			if arr[j] > arr[j+1] {
				fmt.Printf("We have a match %v : %v \n", arr[j], arr[j+1])
				swap(j, j+1)
				fmt.Printf("Collection after swap %v \n", arr)
			}
		}
	}

	fmt.Printf("Collection %v", arr)
}

func swap(left, right int) {
	leftVal := arr[left]
	rightVal := arr[right]
	fmt.Printf("Swapping Values for Left: %v in index %v and Right: %v in index %v \n", leftVal, left, rightVal, right)
	var temp int = arr[left]
	arr[left] = arr[right]
	arr[right] = temp
}

File: ruby/bubble_sort.rb
require_relative 'array_randomizer'

class BubbleSort
  attr_reader :randomized_array

  def initialize
    @randomized_array = ArrayRandomizer.new(10).randomize
  end

  def sort
    pointer = randomized_array.length - 1
    # Iterate Starting From Right index
    while pointer > 1
      # Iterate over elements of array as long as index doesn't overlap with
      # pointer
      randomized_array.each_with_index do |elem, idx|
        # Swap values if left value is greater than right value
        swap(idx, idx + 1) if idx < pointer && elem > randomized_array[idx + 1]
      end
      pointer -= 1 # decrement pointer
    end

    randomized_array
  end

  def swap(left, right)
    temp = randomized_array[left]
    randomized_array[left] = randomized_array[right]
    randomized_array[right] = temp
  end
end
